gemrb/plugins/SDLAudio/SDLAudio.cpp

```diff
@@ -117,52 +113,93 @@ void SDLAudio::music_callback(void *udata, unsigned short *stream, int len) {
 	SDL_mutexV(driver->OurMutex);
 }
 
-void SDLAudio::channel_done_callback(int channel) {
-	SDL_mutexP(g_sdlaudio->OurMutex);
-	assert(g_sdlaudio);
-	assert((unsigned int)channel < g_sdlaudio->channel_data.size());
-	assert(g_sdlaudio->channel_data[channel]);
-	free(g_sdlaudio->channel_data[channel]);
-	g_sdlaudio->channel_data[channel] = NULL;
-	SDL_mutexV(g_sdlaudio->OurMutex);
+bool SDLAudio::evictBuffer()
+{
+	// Note: this function assumes the caller holds bufferMutex
+
+	// Room for optimization: this is O(n^2) in the number of buffers
+	// at the tail that are used. It can be O(n) if LRUCache supports it.
+	unsigned int n = 0;
+	void *p;
+	const char *k;
+	bool res;
+
+	SDL_LockAudio();
+
+	while ((res = buffercache.getLRU(n, k, p)) == true && buffercache.GetCount() >= BUFFER_CACHE_SIZE) {
+		CacheEntry *e = (CacheEntry*)p;
+		bool chunkPlaying = false;
+		int numChannels = Mix_AllocateChannels(-1);
+
+		for (int i = 0; i < numChannels; ++i) {
+			if (Mix_Playing(i) && Mix_GetChunk(i) == e->chunk) {
+				chunkPlaying = true;
+				break;
+			}
+		}
+
+		if (chunkPlaying) {
+			++n;
+		} else {		
+			//Mix_FreeChunk(e->chunk) fails to free anything here
+			free(e->chunk->abuf);
+			free(e->chunk);
+			delete e;
+			buffercache.Remove(k);
+		}
+	}
+
+	SDL_UnlockAudio();
+
+	return res;
 }
 
-Holder<SoundHandle> SDLAudio::Play(const char* ResRef, unsigned int channel,
-	int XPos, int YPos, unsigned int flags, unsigned int *length)
+void SDLAudio::clearBufferCache()
 {
-	// TODO: some panning
-	(void)XPos;
-	(void)YPos;
+	// Room for optimization: any method of iterating over the buffers
+	// would suffice. It doesn't have to be in LRU-order.
+	void *p;
+	const char *k;
+	int n = 0;
+	while (buffercache.getLRU(n, k, p)) {
+		CacheEntry *e = (CacheEntry*)p;
+		free(e->chunk->abuf);
+		free(e->chunk);
+		delete e;
+		buffercache.Remove(k);
+	}
+}
 
-	// TODO: flags
-	(void)flags;
+Mix_Chunk* SDLAudio::loadSound(const char *ResRef, unsigned int &time_length)
+{
+	Mix_Chunk *chunk = nullptr;
+	CacheEntry *e;
+	void *p;
 
-	if (!ResRef) {
-		if (flags & GEM_SND_SPEECH) {
-			Mix_HaltChannel(0);
-		}
-		return Holder<SoundHandle>();
+	if (!ResRef[0]) {
+		return chunk;
+	}
+
+	if(buffercache.Lookup(ResRef, p)) {
```

Comment:
```suggestion
	if (buffercache.Lookup(ResRef, p)) {
```
The style report will take a while, so let's just polish this manually.